okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java

```diff
@@ -0,0 +1,238 @@
+                                if (!redirectTo.isPresent()) {
+                                    callback.onFailure(call, new IOException("Failed to determine valid redirect URL "
+                                            + "for '" + exception.getRedirectTo() + "' and base URLs "
+                                            + urls.getBaseUrls()));
+                                } else {
+                                    Request redirectedRequest = request().newBuilder()
+                                            .url(redirectTo.get())
+                                            .build();
+                                    client.newCallWithMutableState(
+                                            redirectedRequest, backoffStrategy, maxNumRelocations - 1)
+                                            .enqueue(callback);
+                                }
+                            }
+                            return null;
+                        }
+
+                        @Override
+                        public Void visit(QosException.Unavailable exception) {
+                            Optional<Duration> backoff = backoffStrategy.nextBackoff();
+                            if (!backoff.isPresent()) {
+                                log.debug("Max number of retries exceeded, failing call");
+                                callback.onFailure(call, new IOException("Failed to complete the request due to a "
+                                        + "server-side QoS condition: 503", exception));
+                            } else {
+                                log.debug("Rescheduling call after backoff",
+                                        SafeArg.of("backoffMillis", backoff.get().toMillis()));
+                                scheduleExecution(callback, backoff.get());
+                            }
+                            return null;
+                        }
+                    });
+                    return;
+                }
+
+                // Handle responses that correspond to RemoteExceptions / SerializableErrors
+                Optional<RemoteException> httpError = remoteExceptionHandler.handle(response);
+                if (httpError.isPresent()) {
+                    callback.onFailure(call, new IoRemoteException(httpError.get()));
+                    return;
+                }
+
+                // Catch-all: handle all other responses
+                Optional<IOException> ioException = catchAllExceptionHandler.handle(response);
+                if (ioException.isPresent()) {
+                    callback.onFailure(call, ioException.get());
+                    return;
+                }
+
+                callback.onFailure(call, new IOException("Failed to handle request, this is an http-remoting bug."));
+            }
+        });
+    }
+
+    @Override
+    public Response execute() throws IOException {
+        SettableFuture<Response> future = SettableFuture.create();
```

Comment:
Why guava SettableFuture over java8 CompletableFuture?

Reply:
It's mostly a religious question.